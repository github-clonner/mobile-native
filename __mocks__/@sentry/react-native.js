export const captureMessage = jest.fn();
export default {};
