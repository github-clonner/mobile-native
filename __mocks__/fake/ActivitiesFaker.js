
function load(count, container) {

  let activities = [...Array(count)].map((_, i) => {
    const code = 'activityguid' + i;
    return {
      attachment_guid: false,
      blurb: false,
      container_guid:code,
      custom_data:false,
      custom_type:false,
      rowKey: 'something' + i,
      description:"Congratulations! ",
      edited:"",
      guid:code,
      mature:false,
      time_created: "1522036284",
      ownerObj:{
        guid: "824853017709780997",
        type: "user",
        subtype: false,
        time_created: "1522036284",
        // getAvatarSource: () => {
        //   return {
        //     source:'http://thisisaurl'
        //   }
        // }
      },
      shouldBeBlured: jest.fn(),
      message:"Message",
      title:'TITLE',
      owner_guid:"824853017709780997",
      parent_guid:"838106762591510528",
      perma_url:false,
      thumbnail_src:false,
      type:"activity",
      wire_totals: {
        tokens: 1000000000000000000
      },
      containerObj: container
      // _list: {
      //   viewed: {
      //     viewed: new Map([["1019155171608096768",true]]),
      //     addViewed: () => {
      //       return;
      //     }
      //   }
      // },
      // getThumbSource: () => {
      //   return {
      //     source:'http://thisisaurl'
      //   }
      // }
    }
  });

  return {
    'activities': activities,
    'load-next': 'aaaaaa',
    'load-previous': 'aaaaaa'
  }

}

/**
 * comment service data faker
 * @param {integer} count
 */
export function activitiesServiceFaker() {
  return { load: load}
}

