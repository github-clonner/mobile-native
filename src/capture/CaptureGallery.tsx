//@ts-nocheck
import React, { PureComponent } from 'react';

import {
  StyleSheet,
  TouchableOpacity,
  Image,
  Platform,
  FlatList,
} from 'react-native';

import CameraRoll from '@react-native-community/cameraroll';

import Icon from 'react-native-vector-icons/Ionicons';

import CenteredLoading from '../common/components/CenteredLoading';
import androidPermissionsService from '../common/services/android-permissions.service';
import logService from '../common/services/log.service';
import { CommonStyle } from '../styles/Common';
import { View } from 'react-native-animatable';
import ThemedStyles from '../styles/ThemedStyles';
/**
 * Gallery View
 */
export default class CaptureGallery extends PureComponent {
  listRef = null;

  state = {
    photos: [],
    imagesLoaded: false,
    offset: '',
    hasMore: true,
    loading: false,
  };

  static navigationOptions = {
    header: null,
    tabBarIcon: ({ tintColor }) => (
      <Icon name="md-radio-button-on" size={24} color={tintColor} />
    ),
  };

  /**
   * Load Photos
   */
  loadPhotos = async () => {
    let allowed = true;
    if (Platform.OS !== 'ios') {
      allowed = await androidPermissionsService.checkReadExternalStorage();
      if (!allowed) {
        allowed = await androidPermissionsService.readExternalStorage();
      }
    }

    if (allowed === true) {
      this._loadPhotos();
    }
  };

  /**
   * Load photos
   */
  _loadPhotos = async () => {
    if (this.state.loading || !this.state.hasMore) {
      return;
    }

    const params = {
      first: 30,
      assetType: 'All',
    };

    this.setState({ loading: true });

    if (Platform.OS === 'ios') {
      params.groupTypes = 'All';
    }
    if (this.state.offset) {
      params.after = this.state.offset;
    }

    try {
      const result = await CameraRoll.getPhotos(params);

      this.setState({
        imagesLoaded: true,
        photos: this.state.photos.concat(result.edges),
        offset: result.page_info.end_cursor,
        hasMore: result.page_info.has_next_page,
        loading: false,
      });
    } catch (err) {
      logService.exception('[CaptureGallery] loadPhotos', err);
      this.setState({ loading: false });
    }
  };

  /**
   * on component mount load photos
   */
  componentDidMount() {
    setTimeout(() => {
      this.loadPhotos();
    }, 50);
  }

  /**
   * Render
   */
  render() {
    return (
      <View
        style={[
          ThemedStyles.style.backgroundSecondary,
          CommonStyle.flexContainer,
        ]}>
        {this.state.imagesLoaded ? (
          <FlatList
            ref={this.setListRef}
            ListHeaderComponent={this.props.header}
            data={this.state.photos}
            renderItem={this.renderTile}
            style={[
              CommonStyle.flexContainer,
              ThemedStyles.style.backgroundSecondary,
            ]}
            onEndReached={this._loadPhotos}
            ListFooterComponent={
              this.state.loading ? <CenteredLoading /> : null
            }
            numColumns={3}
          />
        ) : (
          <CenteredLoading />
        )}
      </View>
    );
  }

  /**
   * Sets List reference
   */
  setListRef = (ref) => (this.listRef = ref);

  /**
   * render list tile
   */
  renderTile = (item) => {
    const node = item.item.node;

    const icon = node.type.startsWith('video') ? (
      <View style={[CommonStyle.positionAbsolute, CommonStyle.centered]}>
        <Icon name="ios-play-circle" size={24} style={CommonStyle.colorWhite} />
      </View>
    ) : null;

    return (
      <TouchableOpacity
        style={styles.tileImage}
        key={item.index}
        onPress={() => {
          this.onSelected({
            uri: node.image.uri,
            type: node.type,
            fileName: node.image.filename,
            duration: node.image.playableDuration,
            width: node.image.width,
            height: node.image.height,
          });
        }}
        testID={`GalleryImage${item.index}`}>
        <Image
          source={{ uri: node.image.uri }}
          style={styles.tileImage}></Image>
        {icon}
      </TouchableOpacity>
    );
  };

  /**
   * On media selected
   */
  onSelected = (response) => {
    // scroll to top on selection
    this.listRef.scrollToOffset({ x: 0, y: 0, animated: true });

    this.props.onSelected(response);
  };
}

const styles = StyleSheet.create({
  tileImage: {
    minHeight: 120,
    flex: 1,
    padding: 1,
  },
});
