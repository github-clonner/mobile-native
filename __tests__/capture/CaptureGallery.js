import 'react-native';
import React from 'react';
import { Platform, TouchableOpacity } from 'react-native';
import { shallow, render } from 'enzyme';
import CameraRoll from '@react-native-community/cameraroll';

import androidPermission from '../../src/common/services/android-permissions.service';
import { getPhotosFaker } from '../../__mocks__/fake/CameraRollFaker';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('../../src/common/services/android-permissions.service');
jest.mock('@react-native-community/cameraroll');
CameraRoll.getPhotos = jest.fn();
// fake camera roll data
const response = getPhotosFaker(5);
CameraRoll.getPhotos.mockResolvedValue(response);
import CaptureGallery from '../../src/capture/CaptureGallery';

/**
 * Tests
 */
describe('cature gallery component', () => {
  beforeEach(() => {
    CameraRoll.getPhotos.mockClear();
    androidPermission.checkReadExternalStorage.mockClear();
    androidPermission.readExternalStorage.mockClear();
  });

  it('should renders correctly', async () => {
    const galley = renderer.create(<CaptureGallery />).toJSON();
    expect(galley).toMatchSnapshot();
  });

  it('should load photos on mount', () => {
    Platform.OS = 'ios';

    const wrapper = shallow(<CaptureGallery />);
    const instance = wrapper.instance();
    const spyWillMount = jest.spyOn(instance, 'loadPhotos');
    jest.advanceTimersByTime(1000);
    instance.componentDidMount();
    expect(spyWillMount).toHaveBeenCalled();
  });

  it('should check permissions on android before load photos', async () => {
    const checkReadExternalStorage = androidPermission.checkReadExternalStorage;
    const readExternalStorage = androidPermission.readExternalStorage;

    Platform.OS = 'android';

    const wrapper = shallow(<CaptureGallery />);

    // load phoyos
    await wrapper.instance().loadPhotos();

    expect(checkReadExternalStorage).toHaveBeenCalled();
  });

  it('should ask for permissions on android if necessary', async () => {
    const checkReadExternalStorage = androidPermission.checkReadExternalStorage;
    const readExternalStorage = androidPermission.readExternalStorage;

    checkReadExternalStorage.mockResolvedValue(false);

    Platform.OS = 'android';

    const wrapper = shallow(<CaptureGallery />);

    // load phoyos
    await wrapper.instance().loadPhotos();

    expect(readExternalStorage).toHaveBeenCalled();
  });

  it('should calls onSelected when the user select an image', async (done) => {
    const mockFn = jest.fn();

    try {
      const wrapper = renderer.create(<CaptureGallery onSelected={mockFn} />);

      // load phoyos
      await wrapper.getInstance()._loadPhotos();

      expect(CameraRoll.getPhotos).toBeCalled();

      // find TouchableOpacity (rendered images in lists)
      const images = wrapper.root.findAllByType(TouchableOpacity);

      images[0].props.onPress();

      // expect fn to be called once
      expect(mockFn).toBeCalled();
      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should show loaded images', async (done) => {
    const mockFn = jest.fn();

    try {
      const wrapper = renderer.create(<CaptureGallery onSelected={mockFn} />);

      // load phoyos
      await wrapper.getInstance()._loadPhotos();

      // find TouchableOpacity (rendered images in lists)
      const images = wrapper.root.findAllByType(TouchableOpacity);

      // expect 5 images rendered
      expect(images.length).toEqual(5);
      done();
    } catch (e) {
      done.fail(e);
    }
  });
});
