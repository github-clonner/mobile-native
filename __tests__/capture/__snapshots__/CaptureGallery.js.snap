// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cature gallery component should renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "flex": 1,
        "justifyContent": "center",
        "minHeight": 100,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;
